image_builder/image_builder/docker/builder.py

```diff
@@ -40,27 +39,44 @@ def build(
             plugin_engine: PluginEngine,
     ) -> Tuple[str, str, Optional[str]]:
         """Build image from manifest file in a workspace directory and return built image name"""
-        job_templates = _gather_job_templates(config, plugin_engine)
-        assert manifest.lang in job_templates, f'language {manifest.lang} is not supported, supported are {list(job_templates.keys())}'
+        job_templates = _gather_job_templates(plugin_engine)
+        assert job_templates, f'language {manifest.lang} is not supported. Extend Racetrack with job type plugins'
+        assert manifest.lang in job_templates, f'language {manifest.lang} is not supported, supported are: {list(job_templates.keys())}'
 
-        base_image, template_path = job_templates[manifest.lang]
+        base_image_path, template_path, job_type_version = job_templates[manifest.lang]
 
         _wait_for_docker_engine_ready()
 
-        dockerfile_path = workspace / '.fatman.Dockerfile'
-        with wrap_context('templating Dockerfile'):
-            racetrack_version = os.environ.get('DOCKER_TAG', 'latest')
-            template_dockerfile(manifest, template_path, dockerfile_path, base_image,
-                                git_version, racetrack_version, env_vars)
-
-        full_image = get_fatman_image(config.docker_registry, config.docker_registry_namespace, manifest.name, tag)
         metric_labels = {
             'fatman_name': manifest.name,
             'fatman_version': manifest.version,
         }
         Path(config.build_logs_dir).mkdir(parents=True, exist_ok=True)
         logs_filename = f'{config.build_logs_dir}/{deployment_id}.log'
 
+        with wrap_context('building base image'):
+            base_image = join_paths(config.docker_registry, config.docker_registry_namespace, 'fatman-base', f'{manifest.lang}:{job_type_version}')
+            if not _image_exists_in_registry(base_image):
+                base_logs_filename = f'{config.build_logs_dir}/{deployment_id}.base.log'
+                logger.info(f'base image not found in a registry, rebuilding {base_image}, '
+                            f'deployment ID: {deployment_id}, keeping logs in {base_logs_filename}')
+                build_container_image(
+                    base_image,
+                    base_image_path,
+                    base_image_path.parent,
+                    metric_labels,
+                    base_logs_filename,
+                )
+                logger.info(f'base Fatman image has been built and pushed: {base_image}')
```

Comment:
Nitpick: there's quite a lof happening in this function `build`, here and below. Perhaps 1-2 parts could be extracted to separate functions (only if it makes sense).